eventsourcing/src/main/java/org/axonframework/eventsourcing/AnnotationBasedEventStateApplier.java

```diff
@@ -0,0 +1,136 @@
+        this(modelType, ClasspathParameterResolverFactory.forClass(modelType), messageTypeResolver);
+    }
+
+    /**
+     * Initialize a new {@link AnnotationBasedEventStateApplier}.
+     *
+     * @param modelType           The type of model this instance will handle state changes for.
+     * @param parameterResolverFactory the strategy for resolving handler method parameter values
+     * @param messageTypeResolver      The {@link MessageTypeResolver} resolving the
+     *                                 {@link org.axonframework.messaging.MessageType types} for
+     *                                 {@link org.axonframework.eventhandling.EventMessage EventMessages}
+     */
+    private AnnotationBasedEventStateApplier(
+            @Nonnull Class<M> modelType,
+            @Nonnull ParameterResolverFactory parameterResolverFactory,
+            @Nonnull MessageTypeResolver messageTypeResolver) {
+        this(parameterResolverFactory,
+             ClasspathHandlerDefinition.forClass(modelType),
+             messageTypeResolver);
+    }
+
+    /**
+     * Initialize a new {@link AnnotationBasedEventStateApplier}.
+     *
+     * @param parameterResolverFactory the strategy for resolving handler method parameter values
+     * @param handlerDefinition        the handler definition used to create concrete handlers
+     * @param messageTypeResolver      The {@link MessageTypeResolver} resolving the
+     *                                 {@link org.axonframework.messaging.MessageType types} for
+     *                                 {@link org.axonframework.eventhandling.EventMessage EventMessages}
+     */
+    public AnnotationBasedEventStateApplier(@Nonnull ParameterResolverFactory parameterResolverFactory,
+                                            @Nonnull HandlerDefinition handlerDefinition,
+                                            @Nonnull MessageTypeResolver messageTypeResolver) {
+        requireNonNull(parameterResolverFactory, "ParameterResolverFactory may not be null");
+        requireNonNull(handlerDefinition, "HandlerDefinition may not be null");
+        requireNonNull(messageTypeResolver, "MessageTypeResolver may not be null");
+        this.parameterResolverFactory = parameterResolverFactory;
+        this.handlerDefinition = handlerDefinition;
+        this.messageTypeResolver = messageTypeResolver;
+    }
+
+    @Override
+    public M apply(M model, @Nonnull EventMessage<?> event) {
+        requireNonNull(model, "Model may not be null");
+        requireNonNull(event, "Event Message may not be null");
+
+        // todo: should we create a new instance of AnnotationEventHandlerAdapter for each event?
+        // any idea how to cache that?
+        var handlerAdapter = new AnnotationEventHandlerAdapter(
+                model,
+                parameterResolverFactory,
+                handlerDefinition,
+                messageTypeResolver
+        );
```

Comment:
I'm not sure that It's good to create this object on every invocation.
Do you have any idea how to cache that? Model type has no bounds, for currently it could be anything (difficult to find even key for cache).

Reply:
I think one option might be to extract some logic from `AnnotationEventHandlerAdapter` which is based just on the type of the handler (model), not particular instance and cache that in internal map. 
I mean part like:
```
        Optional<MessageHandlingMember<? super Object>> handler =
                inspector.getHandlers(listenerType)
                         .filter(h -> h.canHandle(event, null))
                         .findFirst();
```

Reply:
Do you mean create some `AnnotationBasedEventHandlingComponent` using the inspector related logic from `AnnotationEventHandlerAdapter` and then use `AnnotationBasedEventHandlingComponent` for certain model insie/as a parameter of `AnnotationBasedEventStateApplier` or even the `AnnotationBasedEventStateApplier` will extends the `AnnotationBasedEventHandlingComponent` itself?

Reply:
I think extracting inspector it's OK, but using EventHandlingComponent not exactly. For the MVP it's OK to support just mutating state in the EventStateApplier, but I hope in the future we'd be able to return new objects from `@EventSourcingHandler`, which I cannot achieve with `EventHandlingComponent`, because it returns MessageStream.Empty.

Reply:
I'm not sure if I can use `EventHandlingComponent` directly inside, because it doesn't allow me to return something `NoMessage`, so I cannot return a new state and support record for example. I've done it extracting some code, which will be probably duplicated in the `AnnotatedEventHandlingComponent`.

Reply:
Oh, but if We adjust all the handler to the new API it wouldn't be possible either :/  Hmm.... I'm not sure about the NoMessage as a EventHandlingComponent return type.

Reply:
Discussed with Steven: keep this PR smaller and implement `AnnotatedEventHandlingComponent` in the next one (maybe use it in the EventStateApplier eventually).